Replumbing: make it possible for providers to specify multiple names

levitte opened the pull request:
This modifies OSSL_ALGORITHM to support a NULL-terminated list of
names instead of just one name.

Related to #8967

-----

NOTE: This includes #9969, please review that one first

levitte committed:
Replumbing: make it possible for providers to specify multiple names

This modifies OSSL_ALGORITHM to support a NULL-terminated list of
names instead of just one name.

Related to #8967

levitte committed:
Adapt existing providers to have name lists

paulidale commented:
@slontis and I discussed this recently.  I'm not sure this is a great path to follow -- making all names into arrays of names seems like it is needlessly complex.  Using a single name should be the norm.

We were thinking along the lines of using the object database so that the core can populate aliases automatically (i.e. if the name matches either SN_XXX or LN_XXX, add both).  The core needs the object database regardless, so there is no waste there and it allows providers to not have to worry with aliasing.

Well almost, there also needs to be a way for a provider to add aliases -- I'm thinking an explicit function that adds an alias (or aliases) for an already defined name.

levitte commented:
Contrary to you, I claim that the new core has zero need for the old method object database. The namemap replaces it.

Something I also realised is also that we shouldn't copy the name string when creating new method instances, but rather the namemap id. That means that we can easily support multiple names there without having to duplicate a whole darn array.

levitte commented:
Regarding pre-populating the namemap, see #8984

paulidale commented:
OIDs still depend on the object database....

levitte commented:
Yeah, but the new core doesn't use OIDs, at all. It just uses names.
We will still need to do OID to name translation and therefore have a OID\<-\>name db, but that's a separate thing. (and sure, we can add OID support, and probably should, *but we don't have that right now*)

paulidale commented:
What about restricting each algorithm to a single name and having the core know about some aliases for backwards compatibility?  It isn't like we need to have multiple names for things, they are handles to access algorithms as far as the core is concerned.

This would let us clean up some of the mess (e.g. _id-XXX_ algorithms).

If a provider really does want the same thing registered under two names, it can register it twice.  It will get different IDs but they're not publicly visible or terribly meaningful (unlike NIDs).

slontis commented:
> What about restricting each algorithm to a single name and having the core know about some aliases for backwards compatibility? It isn't like we need to have multiple names for things, they are handles to access algorithms as far as the core is concerned.

Just about everything is registered with 2 names currently in the old system (SN and LN)

paulidale commented:
The two names are often not distinct when looked up case insensitively.

levitte commented:
> What about restricting each algorithm to a single name and having the core know about some aliases for backwards compatibility? It isn't like we need to have multiple names for things, they are handles to access algorithms as far as the core is concerned.

That's essentially what #8984 does (why not have a look?)

paulidale commented:
I was more thinking of having the core's EVP layer handle the mapping and not duplicating names all over the place in different name maps.  The effect is similar though.

I think we can rely on providers to use the canonical name for each algorithm but not application code.  The former is new and we can define the rules, the latter is legacy and the rules escaped long ago.

levitte commented:
> I was more thinking of having the core's EVP layer handle the mapping and not duplicating names all over the place in different name maps. The effect is similar though.

Ah, so actually *exactly* the same, just different spot...

> I think we can rely on providers to use the canonical name for each algorithm but not application code.

...  I think we need a zoomed out (i.e. not just technical) discussion on this.
What's the "canonical" name?  Who's the authority that decides this?

Just to take an example that we have in our code, scrypt.  According to the [ASN.1 module in RFC 7914](https://tools.ietf.org/html/rfc7914#section-7.1), it goes by "id-scrypt":

```
   id-scrypt OBJECT IDENTIFIER ::= {1 3 6 1 4 1 11591 4 11}
```

*We* would like to call it just "SCRYPT", because that conforms better with the rest of the names we use, right?
So what would be the "canonical" name, "id-scrypt" or "SCRYPT"?  On whose authority?  Yours?  Mine?  RFC 7914's?

So, as a way to *avoid* that kind of dispute, why can't we simply think in terms of a set of names, all equal?

Back to the more technical side of things:
- the new namemap already supports a set of names without making any of them "canonical".  They simply get the same identity number.

A more personal note:
- I loathe the notion of having an implementation in one corner of the code and the names it goes by in another corner.  Having them together in code, like this PR does, avoids having to find the implementation in one spot and then having to go look elsewhere to find the name(s) it goes by.  (Yes, I know that's what we currently do... :-/)

levitte commented:
> > I was more thinking of having the core's EVP layer handle the mapping and not duplicating names all over the place in different name maps. The effect is similar though.
> 
> Ah, so actually *exactly* the same, just different spot...

Oh, wait, I forget...  *#8984 does exactly that!*

levitte commented:
Rebased and cleaned up to match the changes that have been made re name vs name_id.

This *almost* tests right...  only test_sslapi acts up, still.

levitte commented:
Speaking of this, I realised there's really no reason why this PR and #8984 should be alternatives for each other...  they should actually combine perfectly well.

levitte commented:
Ah, figured the last failure out...  I had mixed up certain names with each other.
(some ciphers with different key sizes are in 256, 192, 128 order while others are in 128, 192, 256 order...  when cutting and pasting a bit and getting in a flow, it's no wonder I didn't get it right in some places...)

levitte commented:
This is, btw, no longer WIP

levitte commented:
This includes #9969, please review that one first

paulidale commented:
What I'm see here is an increase in complexity for no real benefit.  Using arrays of names everywhere increases the difficult to use.  I still think a provider should register its algorithm using a single name -- it chooses which.  If it matches a preloaded name, it will pick up an aliases automatically.  In this case both "id-scrypt" and "scrypt" would work.

We definitely will need to allow a provider to add its own aliases in addition, but that would be fine as a separate function call.

levitte commented:
Me, I'm seeing a different complexity in having the names and the associated implementation array spread into different corners of the code.  With what you propose, we'd have pre-populated names in one corner (which we can still do), OSSL_DISPATCH array on another corner, and provider addition of more names in a third.  Having to chase around the code to figure out if you have all the names right is *tedious*

levitte commented:
I have to ask, these separate function calls to add more names, where do you see them happen?  Is it an upcall (provider calling back to libcrypto) or a downcall (libcrypto querying for aliases)?

paulidale commented:
I totally agree that different places for names isn't good.  I think we will need the pre-populated names in libcrypto regardless.  A provider is free to lay out its source code as it wishes -- i.e. the separation between the OSSL_DISPATCH and any aliases isn't necessarily under our control.  However, we might be able to do something to encourage it.

I feel that this PR goes too far -- there are lots of arrays `{ "name", NULL }` which are adding indirection unnecessarily.  I believe that a significant amount of the time, a single name will be sufficient -- I think the code backs this view up to some extent (more so with the libcrypto pre-population).  Could we consider an extra _aliases_ argument for these calls?  I.e. `"name", NULL` or `"name", { "aliases", ... }`?  I'd be willing to support that.

As for upcalls or downcalls (autocorrect changed it to _downfalls_ BTW :), I'm not fussed.  After registration, a downcall might make sense.  Before registration an upcall doesn't.  With a lock held, the provider could cache the downcall results between registration and asking for the aliases.

levitte commented:
> A provider is free to lay out its source code as it wishes...

Yes, of course.  But when it comes to *our* providers, we should care.

> I.e. `"name", NULL` or `"name", { "aliases", ... }`?

So...  for scrypt, you'd get something like this:

``` C
static const OSSL_ALGORITHM deflt_kdfs[] = {
    ...
    { "id-scrypt", kdf_scrypt_aliases, "default=yes", kdf_scrypt_functions },
    ...
};
```

So then we have one name directly here, and the rest of the names somewhere else...  together with the definition of `kdf_scrypt_functions` or close to the definition of `deflt_kdfs`?

Code wise, it's not going to make much difference...  we will have to check for an extra NULL (to see if there are associated extra names or not), and if it isn't NULL, we'll have the exact same loop through the extra names array.  Execution wise, I don't see that we gain a lot, if any at all.  and with this, we have the "main name" in one corner, and the extra names somewhere else...

levitte commented:
> As for upcalls or downcalls (autocorrect changed it to _downfalls_ BTW :), I'm not fussed.  After registration, a downcall might make sense.  Before registration an upcall doesn't.  With a lock held, the provider could cache the downcall results between registration and asking for the aliases.

Speaking of complexity...

levitte commented:
I had a thought that maybe your view and mine are much more radically different than I might have imagined, so answer me this: do you see the process of name registration as *entirely separate* from registration of implementation?  If this is the case, the `name` field in `OSSL_ALGORITHM` would work as a tag for an already existing set of names, rather than a name to be registered, right?

beldmit commented:
> I had a thought that maybe your view and mine are much more radically different than I might have imagined, so answer me this: do you see the process of name registration as _entirely separate_ from registration of implementation? If this is the case, the `name` field in `OSSL_ALGORITHM` would work as a tag for an already existing set of names, rather than a name to be registered, right?

My 2c.
Currently, these processes are almost entirely separated. In the ideal world, they should be united in this or that form (BTW, I still don't get how we deal with building ciphersuites from the provider-provided parts).

levitte commented:
> (BTW, I still don't get how we deal with building ciphersuites from the provider-provided parts).

As far as I know, we haven't dealt with that at all yet. However, I'm currently thinking that we could replace any NID in our internal tables with names,and let fetching do the rest. 
(I haven't looked at those tables for a long time, so I may misremember what they look like) 

As for building those tables *dynamically*, I don't think we have *any* such support... and I can't recall that we have made plans for that for 3.0. However, PR welcome, should anyone take an interest...

beldmit commented:
> > (BTW, I still don't get how we deal with building ciphersuites from the provider-provided parts).
> 
> As far as I know, we haven't dealt with that at all yet. However, I'm currently thinking that we could replace any NID in our internal tables with names,and let fetching do the rest.
> (I haven't looked at those tables for a long time, so I may misremember what they look like)

Well, it anyway means we have to hardcode the "standard" names in some file similar to current `objects.txt`.

paulidale commented:
I think that alias registration is a separate process.
We do need to register with a name.

TLS cipher suites are problematic, I'm not sure how to deal with them in the provider & property world.  Ideally, a provider should be able to add new suites.

levitte commented:
> I think that alias registration is a separate process.
> We do need to register with a name.

Still about that "main name" and "aliases" separation? Wanna tell me why that's so important, especially since it's really just surface fluff (the namemap supports a set of names tied to a number, there is zero "main name" support in there... if you really want something "main" in there, it's that number, and all names are aliases for it)

levitte commented:
> > > (BTW, I still don't get how we deal with building ciphersuites from the provider-provided parts).
> > 
> > As far as I know, we haven't dealt with that at all yet. However, I'm currently thinking that we could replace any NID in our internal tables with names,and let fetching do the rest.
> > (I haven't looked at those tables for a long time, so I may misremember what they look like)
> 
> Well, it anyway means we have to hardcode the "standard" names in some file similar to current `objects.txt`.

Not entirely true, we don't really *have* to.  But I agree that it's practical if there's some kind of registry of names that OpenSSL uses internally and therefore expects to be able to use.  We have a few such names already in `include/openssl/core_names.h`, is that suitable enough?

Either way, we can't escape have some kind of connector between what libcrypto / libssl wants to be able to use and what the providers provides.  We have agreed on names.  However, it's quite true that if providers could provide TLS cipher suites as well, that would make this problem go away, or at least have a better chance.

So basically, we need a mechanism to add to those tables that are currently constant in `ssl/t1_lib.c` and similar.  That, however, should probably be discussed somewhere else.

paulidale commented:
It was mostly a thought to avoid the additional complexity in this PR.

levitte commented:
Look, I can probably craft a PR that handles name registration separately (but still keep auto-registration of the `OSSL_ALGORITHM` name if need be, exactly like we have right now).  I think it's going to be easier to discuss when we see some actual code instead of having each our own fluffy view of what's more complex than the other.

Will happen tonight or tomorrow...

levitte committed:
squash! Replumbing: make it possible for providers to specify multiple names

New narrative:

This modifies the treatment of algorithm name strings to allow
multiple names separated with colons.

levitte committed:
squash! Adapt existing providers to have name lists

*possibly* have name lists

levitte committed:
fixup! Replumbing: make it possible for providers to specify multiple names

levitte commented:
I've changed this to allow the name strings in `OSSL_ALGORITHM` to contain multiple names.
This does have some drawbacks, such as the namemap having to duplicate the string in, er, surprising places.  Not a biggie, but was a bit of a hurdle 'til I got it right.

Anyway, time for a new review.  The impact in `providers/` is substantially smaller...

levitte commented:
I comment on this implementation: I could have had the parsing of the names string in the namemap functions.  However, the multiple name thing is currently all about algorithms, there's nothing else planned or that I can foresee where central support of multiple names would be a benefit, as other APIs may not want to support that.
(point in case, the `OSSL_STORE` API uses URI schemes as "names", and there's no reason whatsoever to have multiple names for the same scheme...  that's simply not done, as far as I know)

So the multiple name mechanism is confined to EVP.

paulidale commented on crypto/evp/evp_fetch.c:
```diff
@@ -51,6 +53,67 @@ struct method_data_st {
+     * Check that no name is an empty string, and that all names have at
+     * most one numeric identity together.
+     */
+    for (p = names; *p != '\0'; p = q == NULL ? p + l : q + 1) {
```
I recommend some parenthesis in the `p = q == NULL ....` to make it clearer.

levitte replied:
Sure. Only around the conditional, or around the whole ?: expression?

paulidale replied:
I'd do it: `p = (q == NULL ? p + l : q + 1)`

Since I got it in the wrong reply.

levitte replied:
Makes sense. Thanks

levitte replied:
Done

paulidale commented on crypto/evp/evp_fetch.c:
```diff
@@ -51,6 +53,67 @@ struct method_data_st {
+    }
+
+    /* Now that we have checked, register all names */
+    for (p = names; *p != '\0'; p = q == NULL ? p + l : q + 1) {
```
Likewise.

paulidale replied:
I'd do it: `p = (q == NULL ? p + l : q + 1)`

levitte replied:
Makes sense. Thanks

levitte replied:
Done

paulidale commented on crypto/evp/evp_fetch.c:
```diff
@@ -51,6 +53,67 @@ struct method_data_st {
+        this_id = ossl_namemap_add_n(namemap, id, p, l);
+        if (id == 0)
+            id = this_id;
+        if (this_id != id) {
```
`else` before the if?  Like above?

levitte replied:
Sure

levitte replied:
Done

paulidale reviewed (commented):
Is it worthwhile investing in a utility function to convert a delimited list into a stack of strings?
It would avoid some of the mucking around here.

We'd also not need to free the strings since they are being copied into the name map, but that's an optimisation.

levitte committed:
fixup! Replumbing: make it possible for providers to specify multiple names

levitte commented:
Ping

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -51,6 +53,67 @@ struct method_data_st {
+        this_id = ossl_namemap_name2num_n(namemap, p, l);
+
+        if (*p == '\0' || *p == NAME_SEPARATOR) {
+            ERR_raise(ERR_LIB_EVP, ERR_R_INTERNAL_ERROR);
```
Is this actually an *internal* error? Or could this be an error in the data from a provider?

levitte replied:
I had no imagination.  Can you come up with a good reason?  I'm all ears.

mattcaswell replied:
ERR_R_BAD_ALGORITHM_NAME?

levitte replied:
Sure

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -51,6 +53,67 @@ struct method_data_st {
+        if (id == 0)
+            id = this_id;
+        else if (this_id != 0 && this_id != id) {
+            ERR_raise(ERR_LIB_EVP, ERR_R_INTERNAL_ERROR);
```
Again - could this be an error from a provider, i.e. not an internal error? Or in fact inconsistent data from two different providers? This actually implies that providers must not have algorithm synonyms as two separate entries in an OSSL_ALGORITHM array, e.g. this is an error:

````
    { "dhKeyAgreement", "default=yes", dh_keyexch_functions },
    { "dh", "default=yes", dh_keyexch_functions },
````
Because, if another provider defines it like this it will fail:
````
    { "dh:dhKeyAgreement", "default=yes", dh_keyexch_functions },
````

Feasibly, we could actually add a check from such synonymous entries?

levitte replied:
That...  could be done, sure.  So question is, do we want to allow that form of synonym or disallow it?

mattcaswell replied:
I tend towards disallow, to encourage a consistent way of doing things across all providers.

levitte replied:
Hmmm, that's going to take an effort that doesn't quite belong here.  Do you mind if it happens in a separate PR?

mattcaswell replied:
Separate PR is fine.

mattcaswell commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -49,6 +55,12 @@ names already associated with that number.
 I<name>.
 
+ossl_namemap_add_n() and ossl_namemap_name2num_n() do the same thing
+as ossl_namemap_add() and ossl_namemap_name2num(), but take an string
```
Typo: take *a* string

levitte replied:
Fixing

mattcaswell commented on providers/default/defltprov.c:
```diff
@@ -121,19 +121,18 @@ static const OSSL_ALGORITHM deflt_ciphers[] = {
+    { "id-aes256-CCM:AES-256-CCM", "fips=yes", aes256ccm_functions },
+    { "id-aes192-CCM:AES-192-CCM", "fips=yes", aes192ccm_functions },
+    { "id-aes128-CCM:AES-128-CCM", "fips=yes", aes128ccm_functions },
+    { "id-aes256-wrap:AES-256-WRAP:AES256-WRAP", "fips=yes", aes256wrap_functions },
```
Why do we allow "AES-256" and "AES256" here - but not with other algorithms?

levitte replied:
That's a question I cannot answer...  I know that these names exist in previous OpenSSL versions:

```
$ openssl list -cipher-algorithms | grep -i aes | grep -i wrap
aes128-wrap => id-aes128-wrap
aes192-wrap => id-aes192-wrap
aes256-wrap => id-aes256-wrap
id-aes128-wrap
id-aes128-wrap-pad
id-aes192-wrap
id-aes192-wrap-pad
id-aes256-wrap
id-aes256-wrap-pad
```

The `id-` names are from the ASN.1 module, and `aesNNN-wrap` is our "friendlier" name that only half compliant with our other names, so I added `aes-NNN-wrap` to be aligned with our usual *cipher*-*bits*-*mode* naming.

levitte replied:
So basically, `AES256` is the odd name that I haven't replicated because it's relatively uncommon on our naming.

Is this a matter that should be resolved here, or in a separate naming PR?

mattcaswell replied:
So the "AES256-WRAP" name is newly introduced in master? If so, I would drop it in favour of "AES-256-WRAP" so we can have a consistent format for all "friendly" names (and I would put that name first).

levitte replied:
No, "AES-256-WRAP" is newly introduced in this PR.

I forgot a detail:

```
: ; openssl version
OpenSSL 1.1.1d  10 Sep 2019
```

mattcaswell commented on providers/default/defltprov.c:
```diff
@@ -144,9 +143,9 @@ static const OSSL_ALGORITHM deflt_ciphers[] = {
-    { "ARIA-256-CBC", "default=yes", aria256cbc_functions },
-    { "ARIA-192-CBC", "default=yes", aria192cbc_functions },
-    { "ARIA-128-CBC", "default=yes", aria128cbc_functions },
+    { "ARIA-256-CBC:ARIA256", "default=yes", aria256cbc_functions },
```
Why this alias?

levitte replied:
Because it's already there:

```
: ; openssl list -cipher-algorithms|grep -i aria
...
aria128 => ARIA-128-CBC
aria192 => ARIA-192-CBC
aria256 => ARIA-256-CBC
```

mattcaswell replied:
Ok. Weird - but ok.

levitte replied:
There's a *lot* of weirdness in the naming scheme.  Have you read through `crypto/objects/objects.txt`?  It's an interesting read, and even more so when you realise that sometime through history, ASN.1 modules started using that `id-` prefix we all love so much...  this has also affected our naming...  and then we have TLS cipher suite naming vs naming in ASN.1 modules, where one has things like "AES256" while the other has things like "AES-256".

Let's just say that it's a tad messy, and we have adapted to all, or partly so, all depending who implemented the affected algorithm.

paulidale replied:
This one would be my fault and due to my ignorance at the time.

mattcaswell commented on providers/default/defltprov.c:
```diff
@@ -272,21 +271,21 @@ static const OSSL_ALGORITHM deflt_kdfs[] = {
 static const OSSL_ALGORITHM deflt_signature[] = {
 #ifndef OPENSSL_NO_DSA
-    { "DSA", "default=yes", dsa_signature_functions },
+    { "dsaEncryption:DSA", "default=yes", dsa_signature_functions },
```
I know it makes no difference technically. But I'd suggest we adopt a convention of having a preferred "canonical" name as being the first one in the list. It would be this name that we refer to throughout the documentation for example.

levitte replied:
Considering we used to have NIDs all over the place, it's not at all easy to know what the "canonical" name is.  But as long as we agree, I'm game.
(I personally don't care, precisely because it doesn't matter technically)

mattcaswell replied:
I think we ignore any "old" names that we have and pick a naming scheme for our "friendly" names and use that. If it happens to coincide with the any of the old names, then great.

levitte replied:
There's a problem with that...  if someone uses something like `EVP_PKEY_DSA`, which is an alias for a NID, and someone else uses either `OBJ_nid2sn` or `OBJ_nid2ln` on it an uses that name in a fetch, there are going to be some misses unless we make sure that our providers give the full set of names.  Incidently, this is something we need to be careful with any time we do implicit fetches.

mattcaswell replied:
I think you misunderstand my suggestion. I only meant "ignore any old names" in terms of picking our friendly name. Choose a scheme for our friendly names forgetting about what we've called things before. Make that the first listed name. After that list all the aliases that we've historically called an algorithm by. We should encourage use of the "friendly name" from hereon in.

levitte replied:
Ah, right, that makes a lot more sense.

So now, we only have to select the friendly name.  Wanna do the honors and go through the list and suggest which one should be the friendly name (I really don't mind if you do that as bulk suggestion where applicable)

mattcaswell replied:
My suggestion for a naming scheme:

[ALGNAME][VERSION?][-SUBNAME[VERSION?]?][-SIZE?][-MODE?]

Although we're case insensitive, we should still present all names in caps in the "canonical" form.
Subname may be present where we are combining multiple algorithms together, e.g. MD5-SHA1
Version is only present if there are multiple versions of an alg (MD2, MD4, MD5). It may be omitted if there is only one version (if in the future a subsequent version is released we can always change the canonical name, and add the old name as an alias)
Size is only present if multiple versions of an algorithm exist with different sizes (e.g. AES-128-CBC, AES-256-CBC)
Mode is only present where applicable.

Examples:
SHA1
SHA2-256
SHA3-224
BLAKE2S-256
KECCAK-KMAC-128
ARIA-256-CBC
DH
DSA

levitte replied:
Cool, will do

levitte replied:
Ho boy, naming can be fun...  I was wondering about "SHA2-256", 'cause I couldn't recognise it, and no wonder, I actually think you just invented *yet another* set of names for the SHA2 family.  NIST uses SHA-1, SHA-224, SHA-256,SHA-384, SHA-512, SHA-512/224 and SHA-512/256 (see https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf)

So question is, if we want some *real world* names, must we do "SHA2-xxx"?

paulidale replied:
The SHA2's would be algname-size wouldn't they? (i.e. conforming to the naming scheme)
SHA-1 is the odd one out.

mattcaswell replied:
Well we have SHA1 and SHA3, so it seems odd that SHA2 doesn't have the version number included. Personally I'm happy with calling it SHA2-256 etc, with SHA-256 as an alias. It makes it clearer - even though it doesn't entirely conform to NIST naming. But if that is considered a step too far then we can modify my suggested naming scheme to include some additional weasel words about when the version number can be optional. E.g.
 
"Version is only present if there are multiple versions of an alg (MD2, MD4, MD5). It may be omitted if there is only one version (if in the future a subsequent version is released we can always change the canonical name, and add the old name as an alias), or if common practice for a particular version is to omit it"

paulidale replied:
Using SHA2-256 seems sensible, with SHA-256 as an alias.

levitte replied:
I can go for SHA2-nnn, no problem.  I'm just amused that we end up inventing new names for this. :wink:

levitte replied:
Done

mattcaswell reviewed (changes requested):
I have a worry that has occurred to me after reviewing this PR. Does this effectively require all providers to always list all known aliases for each algorithm they implement? The problem I see is whether an application works or not suddenly depends on what providers might be loaded even if those providers are not being used in a particular fetch.

For example, say the default provider defines two aliases for dh: "dh" and "dhKeyAgreeement".
A third party provider defines it's own implementation of "dh" and just calls it "dh".

With the default and 3rd party providers loaded, an application does this:

EVP_KEYEXCH_fetch(NULL, "dhKeyAgreement", "thirdparty=yes");

This works because the default provider is loaded and that defines "dh" and "dhKeyAgreement" as being synonymous.

In a different configuration the same application is run with only the third party provider loaded. Suddenly the same fetch no longer works because we don't know that "dh" and "dhKeyAgreement" are synonyms. So it was the presence of the default provider that made it work before, even though that provider is not being used directly.

levitte commented:
> Does this effectively require all providers to always list all known aliases for each algorithm they implement?

Not really, and it completely depends on the application, doesn't it?  Do note that #8984 pre-loads the namemap with names that *we* know, and it's not in conflict with this one, so if that takes away your worries, that's cool.

But essentially, there are two options:
1. any provider must make sure to include the common names.
2. the application must know at least one of the name that the desire provider uses.

Do note that if there are several providers implementing the same algo but with different names, it's enough that they have at least one name in common, then the *union* of all their names will be usable to reach either.

> The problem I see is whether an application works or not suddenly depends on what providers might be loaded even if those providers are not being used in a particular fetch.

That's a problem even with single name implementations.

> In a different configuration the same application is run with only the third party provider loaded. Suddenly the same fetch no longer works because we don't know that "dh" and "dhKeyAgreement" are synonyms. So it was the presence of the default provider that made it work before, even though that provider is not being used directly.

"know your provider" should be the common answer to that.  #8984 should help a bit, but that only takes care of synonyms that we *do* know, it doesn't take care of stuff we don't (and I don't see a need for us to, either).

levitte committed:
fixup! Replumbing: make it possible for providers to specify multiple names

levitte committed:
fixup! Replumbing: make it possible for providers to specify multiple names

levitte commented:
Rebased the lot too

paulidale commented:
@mattcaswell, compatible naming is main reason I wanted the core to add the aliases it knows about.  #8984 will achieve this.  I wanted to write _consistent naming_, but it isn't currently.

As for providers listing several names for an algorithm, I see this as unnecessary if the core knows them.  One should be enough.  This has been a sticking point between me and @levitte :grinning:

We have to trust that the providers will do the right thing (whatever that is).  A typo in a provider's name list for an algorithm could cause problems like the one suggested -- there is no nice mitigation.  The best I can think of is to: insist that the names must either be all unknown or all known and referring to the same algorithm.  i.e. no additional new names are permitted for algorithms the core knows about.

levitte commented:
I think you're making a problem where there isn't one, or rather, you're making it our problem when it's really somebody else's.  And most of all, it's a social problem, not a technical one.
The basic message here is "know your providers" (for anyone who wants to use them), but also "know the expectations" (for the provider writers).

To put it quite plainly, this has two sides:
1.  if someone writes an application, or configures for a provider that doesn't provide algorithms with the names they expect, they should investigate what the configured providers does use.  This is a use case for the `openssl provider` tool and why that tool is important.
2.  if someone writes a provider and wants to implement some well known algorithms, it would be wise of them to use at least one or two well known names, or face a reality that their provider will probably not grow very popular.

Something that I *don't* want to see is us becoming some kind of registry of all names possible, including for whatever upcoming algorithms.  I believe that could become an unnecessary hurdle.  I know that's not quite what you're proposing, @paulidale, but that's a possible consequence that we have already seen (we've become the central registry of NIDs and associated names).

levitte commented:
The configuration problem that's been mentioned is just same as configuring for using the FIPS module only, and having an application that tries to fetch, say, the "sm3" hash algo.  That will fail because it's not available.  The FAQ should be "Did you check that this providers provides that name?"

levitte committed:
fixup! Adapt existing providers to have name lists

paulidale commented:
I'm not worried about malicious providers -- they lie beyond our control.  I'm worried about a provider that, for example, registers `CAMELIA-256-CBC:CAMELLIA256` and the bad name then being saved in a configuration file.

I guess the core of the message is _know your providers_ as @levitte wrote.

My suggestion was probably a bit extreme after all.

levitte commented:
BTW, if we're inventing a naming standard for *our* names, maybe we should document it in some sane place other than a comment in the code.  Please suggest.

levitte committed:
fixup! Adapt existing providers to have name lists

paulidale commented:
How about: `doc/internal/man7/the_naming_of_cats.pod` :)

I agree it should be documented, either as a README or a POD file.  I lean towards the latter.

slontis commented on crypto/evp/evp_fetch.c:
```diff
@@ -105,10 +170,13 @@ static void *get_method_from_store(OPENSSL_CTX *libctx, void *store,
         OSSL_NAMEMAP *namemap = ossl_namemap_stored(libctx);
+        const char *names = methdata->names;
+        const char *q = strchr(names, NAME_SEPARATOR);
+        size_t l = q == NULL ? strlen(names) : (size_t)(q - names);
```
maybe use the same notation as above l = (q == NULL ? ...);

levitte replied:
Ah, yeah, sure

slontis commented on crypto/evp/evp_fetch.c:
```diff
@@ -131,21 +199,30 @@ static void *get_method_from_store(OPENSSL_CTX *libctx, void *store,
+    if (names != NULL) {
+        const char *q = strchr(names, NAME_SEPARATOR);
+
+        l = q == NULL ? strlen(names) : (size_t)(q - names);
```
here also..

slontis commented on providers/default/defltprov.c:
```diff
@@ -50,16 +50,43 @@ static int deflt_get_params(const OSSL_PROVIDER *prov, OSSL_PARAM params[])
+    { "SHA2-256:SHA-256:SHA256", "default=yes", sha256_functions },
+    { "SHA2-384:SHA-384:SHA384", "default=yes", sha384_functions },
+    { "SHA2-512:SHA-512:SHA512", "default=yes", sha512_functions },
+    { "SHA2-512/224:SHA-512/224:SHA512-224", "default=yes",
```
Is eveyone happy with the / here?

t8m replied:
I am OK with it. What would be the alternative?

slontis replied:
Its probably better than '_'
and '-' would probably be confusing.

paulidale replied:
NIST uses `/` to separate the 512 and 224.  I wonder about the final name `SHA512-224`, I assume this is historic (which would be my fault I suspect).

levitte replied:
It's not very hard to check what names we have used...

```
: ; openssl version
OpenSSL 1.1.1d  10 Sep 2019
: ; openssl list -digest-algorithms
RSA-MD4 => MD4
RSA-MD5 => MD5
RSA-RIPEMD160 => RIPEMD160
RSA-SHA1 => SHA1
RSA-SHA1-2 => RSA-SHA1
RSA-SHA224 => SHA224
RSA-SHA256 => SHA256
RSA-SHA3-224 => SHA3-224
RSA-SHA3-256 => SHA3-256
RSA-SHA3-384 => SHA3-384
RSA-SHA3-512 => SHA3-512
RSA-SHA384 => SHA384
RSA-SHA512 => SHA512
RSA-SHA512/224 => SHA512-224
RSA-SHA512/256 => SHA512-256
RSA-SM3 => SM3
BLAKE2b512
BLAKE2s256
id-rsassa-pkcs1-v1_5-with-sha3-224 => SHA3-224
id-rsassa-pkcs1-v1_5-with-sha3-256 => SHA3-256
id-rsassa-pkcs1-v1_5-with-sha3-384 => SHA3-384
id-rsassa-pkcs1-v1_5-with-sha3-512 => SHA3-512
MD4
md4WithRSAEncryption => MD4
MD5
MD5-SHA1
md5WithRSAEncryption => MD5
ripemd => RIPEMD160
RIPEMD160
ripemd160WithRSA => RIPEMD160
rmd160 => RIPEMD160
SHA1
sha1WithRSAEncryption => SHA1
SHA224
sha224WithRSAEncryption => SHA224
SHA256
sha256WithRSAEncryption => SHA256
SHA3-224
SHA3-256
SHA3-384
SHA3-512
SHA384
sha384WithRSAEncryption => SHA384
SHA512
SHA512-224
sha512-224WithRSAEncryption => SHA512-224
SHA512-256
sha512-256WithRSAEncryption => SHA512-256
sha512WithRSAEncryption => SHA512
SHAKE128
SHAKE256
SM3
sm3WithRSAEncryption => SM3
ssl3-md5 => MD5
ssl3-sha1 => SHA1
whirlpool
```

This all comes from `crypto/objects/objects.txt`, so you can look there too.

(it's interesting to see that the RSA aliases use those slashes, so we do have a precedent)

slontis commented on providers/default/defltprov.c:
```diff
@@ -72,12 +99,20 @@ static const OSSL_ALGORITHM deflt_digests[] = {
+     * If we assume that "2b" and "2s" are versions, that pattern
+     * fits with ours.  We also add our historical names.
+     */
+    { "BLAKE2b-256:BLAKE2s256", "default=yes", blake2s256_functions },
```
I assume that it is not the same as BLAKE2B-256 ??

levitte replied:
Oops

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -342,22 +371,27 @@ static const OSSL_ALGORITHM fips_ciphers[] = {
+      aes192wrap_functions },
+    { "AES-128-WRAP:id-aes128-wrap:AES128-WRAP", "fips=yes",
+      aes128wrap_functions },
+    { "AES-256-WRAP-PADid-aes256-wrap-pad::AES256-WRAP-PAD", "fips=yes",
```
oops.. missing seperator ':' which is why some travis tests are failing in fips.
Still not sure that I like these strings being duplicated across providers.
Would defines be any better?

slontis replied:
following 2 strings are also missing the :

slontis replied:
#define OSSL_CIPHER_NAME_AES_256_WRAP_PAD ""AES-256-WRAP-PAD:id-aes256-wrap-pad"

paulidale replied:
I kind of think that the FIPS provider could get away with just defining the names NIST uses.

With #10070 being a reasonable prospect, we'll be reconsidering how this works...

levitte replied:
My original proposal was to have symbols that carried the names, alongside the implementations, and having fipsprov.c, defltprov.c et al simply use those symbols.

#10070 certainly changes things, but not so much the central naming mechanism, so it shouldn't affect this PR.

slontis replied:
what should ChaCha20-Poly1305 look like?

mattcaswell replied:
Its a combination of two algorithms. According to the naming scheme:

ALGNAME[VERSION?][-SUBNAME[VERSION?]?][-SIZE?][-MODE?]

The first algname is CHACHA20, and there is a subname POLY1305. Therefore:

CHACAH20-POLY1305

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -342,22 +371,27 @@ static const OSSL_ALGORITHM fips_ciphers[] = {
+      aes192wrap_functions },
+    { "AES-128-WRAP:id-aes128-wrap:AES128-WRAP", "fips=yes",
+      aes128wrap_functions },
+    { "AES-256-WRAP-PADid-aes256-wrap-pad::AES256-WRAP-PAD", "fips=yes",
```
Not sure what the double seperator before AES256-WRAP-PAD would do :)

levitte replied:
They are run-away sheep!  I've just nudged them back in their fold...

paulidale replied:
@slontis prefers goats to sheep 🤣

slontis replied:
The best of both worlds
https://www.youtube.com/watch?v=N-lFK8iRG60

levitte committed:
fixup! Adapt existing providers to have name lists

levitte committed:
fixup! Replumbing: make it possible for providers to specify multiple names

richsalz commented:
> I kind of think that the FIPS provider could get away with just defining the names NIST uses.

YES!!!

levitte commented:
So, er, what is the NIST name for AES-265 in CBC mode?  [SP 800-38A](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf) doesn't help much, unless you wanna see names like CBC-AES256...

t8m reviewed (approved):
Looks good now

levitte commented:
I'll take that approval.  Should someone want to make name changes, do so in another PR.

levitte commented:
Merged.

695d195bbb81f8ed4027468fe1c480f958e846c8 Replumbing: make it possible for providers to specify multiple names
df553b79419230d698d221919c7ceec68aa8c6c6 Adapt existing providers to posibly have name lists
